// mh-infrared-gateway/MainV2/ToolModel/ExtendView.js
'use strict';

var React = require('react-native');

var {
  AppRegistry,
  StyleSheet,
  Text,
  TouchableHighlight,
  Image,
  View,
  Component,
  TextInput,
  PixelRatio,
  StatusBar,
  TouchableOpacity,
  Platform,
  ScrollView,
  Dimensions,

  DeviceEventEmitter,
} = React;


var Button = require('../../CommonModules/Button');
var TitleButton = require('../../CommonModules/TitleButton');
var screenWidth = Dimensions.get('window').width;
var screenHeight = Dimensions.get('window').height;
import { NativeModules } from 'react-native';
var MHPluginSDK = NativeModules.MHPluginSDK;
const APPBAR_HEIGHT = Platform.OS === 'ios' ? 44 : 56;
const NavigatorBarHeight = (screenHeight==812)?(88):( APPBAR_HEIGHT +20);

 class NormalExtendView extends React.Component {
  constructor(props, context) {
    super(props, context);

    this.state = {
      did: MHPluginSDK.deviceId,
      model: MHPluginSDK.deviceModel,
      apiLevel: MHPluginSDK.apiLevel,
      basePath: MHPluginSDK.basePath,
      requestStatus: false,
      clicked: 'none',
      devMode: MHPluginSDK.devMode,
      cloumnCount: 3,
      totalCount: this.props.totalCount,
    };
  }

  render() {

    return (
      <View style={styles.containerAll} >
  
        {this.renderAllItem()}

      </View>
    );
  }

  renderAllItem() {
  
    var allItemDate = [];
    for(var i =0 ;i<dataSource.length;i++){

      let title = dataSource[i];
      allItemDate.push(

        <View style ={styles.columnItemBgView} >
            <TitleButton style={styles.buttonView}  key ={title+i} onPress ={()=> this.onPressButton(title)} key = {title} title ={title}  titleSize={15} textColor="rgb(11,11,1)" >
            </TitleButton>
        </View>

      );

    }
    return allItemDate;

  }

  onPressButton (event){
    alert(event);

  }

  componentWillMount() {
    
  }

  componentWillUnmount() {

  }
}


const marginWidth = 10;
const itemWidth = (screenWidth -(4 * marginWidth))/3.0;
const itemHeight = itemWidth * 0.4;


const directionHeight = screenWidth*0.8;
var styles = StyleSheet.create({
  containerAll:{
    flex:1,
    backgroundColor:'#ffffff',
    flexDirection:'row',
    flexWrap:'wrap',
  },
  titleBottom: {
    width:40,
    height:40,
    position:'absolute',
    fontSize: 14,
    marginTop:28,
    alignSelf: 'center',

    justifyContent: 'center', 
    alignItems:'center' 
  },
  contentView:{

  },
  buttonView:{
    width:itemWidth,
    height:itemHeight,
    borderWidth:1,
    borderColor:'#C0C0C0',
    borderRadius:3,
  },
  columnItemBgView:{
    width:itemWidth,
    height:itemHeight,
    backgroundColor:'#ffffff',
    marginTop:marginWidth,
    marginLeft:marginWidth,
    justifyContent:'center',
    alignItems:'center',
    alignSelf:'center',
  }

});

var dataSource = ["空调1","空调2","空调3","空调4","空调5","空调6","空调7","空调8",];


module.exports = NormalExtendView;
