// mh-infrared-gateway/MainV2/ToolModel/NavigatorView.js
var React = require('react-native');
var {
  Component,
  StyleSheet,
  Dimensions,
  TouchableHighlight,
  View,
  Image,
  Text,
  Modal,
  Platform,
} = React;

import { NativeModules } from 'react-native';
var MHPluginSDK = NativeModules.MHPluginSDK;
var screenWidth = Dimensions.get('window').width;
var screenHeight = Dimensions.get('window').height;
const APPBAR_HEIGHT = Platform.OS === 'ios' ? 44 : 56;
const NavigatorBarHeight = (screenHeight==812)?(88):( APPBAR_HEIGHT +20);

const containerContentViewHeight =  (screenHeight==812)?(88):( APPBAR_HEIGHT +20);

var imageModel ={
'defaultImage':'NewControllerImage/remote_status_normal.png',
'highlightImage':'NewControllerImage/remote_status_press.png',
};

class NavigatorView extends React.Component{
  constructor(props) {
    super(props);
    this.state={
      defaultShow:imageModel.defaultImage,
      title:this.props.title,
    };
  }
  
  /* DOM */
  render(){

      var isIphoneX = (screenHeight==812)?(true):( false);
      if(isIphoneX ==true){

        return this.render_iphoneX();
      }else{
        return this.render_notIphoneX();
      }
  
  
  }


  render_iphoneX(){
    console.log('-----------------------render_iphoneX()');
    return(
      <View style={styles.container}>
          <View style={styles.iphoneX_contentView} >
            <Text style={{fontSize:19,marginTop:9,color:'rgba(0,0,0,0)',opacity:1}}>{'  '}{this.state.title}</Text>
            <Image style={[styles.image,{paddingLeft:10,marginLeft:15,marginTop:9,width:20,height:20}]}
            source={{isStatic:!MHPluginSDK.devMode, uri:MHPluginSDK.basePath + this.state.defaultShow}}
            />
          </View>
      </View>
    );
  }

  render_notIphoneX(){
    console.log('-----------------------render_notIphoneX()');
    return(
      <View style={styles.container}>
          <View style={styles.containerContentView} >
            <Text style={{fontSize:19,marginTop:9,color:'rgba(0,0,0,0)',opacity:1}}>{'  '}{this.state.title}</Text>
            <Image style={[styles.image,{paddingLeft:10,marginLeft:15,marginTop:9,width:20,height:20}]}
            source={{isStatic:!MHPluginSDK.devMode, uri:MHPluginSDK.basePath + this.state.defaultShow}}
            />
          </View>
      </View>
    );
  }


  
  /* 显示图层*/
  show(){

    var self = this;
    this.toastTimer =  setTimeout(
      () =>  {
        this.setState({defaultShow:imageModel.highlightImage})
      },
      0
    );
    this.toastTimer =  setTimeout(
      () =>  {
        this.setState({defaultShow:imageModel.defaultImage})
      },
      100
    );
  }

}



const iphoneX_navigatorBarHeight = 44;

var styles = StyleSheet.create({
  container:{
    marginTop:0,
    height:NavigatorBarHeight,
    justifyContent:'center',
    alignItems:'center',
    alignSelf:'center',
  },
  containerContentView:{
    marginTop:20,
    height:NavigatorBarHeight-20,
    // width:180,
    // backgroundColor:'rgb(238,238,238)',
    flexDirection:'row',
    justifyContent:'flex-end'
  },
  iphoneX_contentView:{
    marginTop:NavigatorBarHeight-iphoneX_navigatorBarHeight,
    height:iphoneX_navigatorBarHeight,
    // backgroundColor:'rgb(238,238,238)',
    flexDirection:'row',
    justifyContent:'flex-end'
  },
  image:{

  },
  MessageView:{
    position:'absolute',
    width:screenWidth,
    height:screenHeight/3,
    left:0,
    right:0,
    top:screenHeight*(2/3),
    marginBottom:0,
    justifyContent:'center',
  },
 
});

module.exports = NavigatorView;
