// mh-infrared-gateway/MainV2/TopBoxLogic/IPTVBrand/IPTVBrandList.js
'use strict';

var React = require('react-native');
var MHTableView = require('../../../CommonModules/MHTableView');
var ImageButton = require('../../../CommonModules/ImageButton');
var MIRequestServer = require('../../../MainV2/MIRequestServer/MIRequestServer').MIRequestServer;
var MatchTool = require('../../../MainV2/ConstDefine/matchTool').MatchTool;
var NewMatchRemoter = require('../../../MainV2/MatchRemoter/NewMatchRemoter');
var GateWayRemoter = require('../../../MainV2/GateWay/GateWayRemoter');
var GateWayTool = require('../../../MainV2/MIRequestServer/GateWayTool').GateWayTool;
var DeviceFactory = require('../../../MainV2/ConstDefine/DeviceFactory').DeviceFactory;
var MessageToast = require('../../../MainV2/ToolModel/MessageToast');
var New_studyRemoter = require('../../../MainV2/AllStudyRemoter/New_studyRemoter');
var SearchFilter = require('../../../MainV2/BrandList/SearchFilter');
var LocalizedStrings = require('../../../Main/MHLocalizableString.js').string;


var {
    AppRegistry,
    StyleSheet,
    Text,
    ListView,
    View,
    Image,
    Platform,
    ActivityIndicator,
    AlertIOS,
    Dimensions,
    TouchableHighlight,
    Component,
    TextInput,
    PixelRatio,
} = React;
import { NativeModules } from 'react-native';
var MHPluginSDK = NativeModules.MHPluginSDK;
var LetterArray = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'];
var sourceData = [];

/* 用于内部选择品牌id的 功能*/



class IPTVBrandList extends Component {
    constructor(props) {
        super(props);

        var newDataSource = {
            "index": {
                "foregroundColor": "#ffffff",
            },
            "sections": []
        };
        this.state = {
            did: MHPluginSDK.deviceId,
            model: MHPluginSDK.deviceModel,
            apiLevel: MHPluginSDK.apiLevel,
            basePath: MHPluginSDK.basePath,
            requestStatus: false,
            deviceType: this.props.deviceType,
            devMode: MHPluginSDK.desvMode,
            dataSource: newDataSource,
            select_GateWayId: this.props.select_GateWayId,
            device_Factory: new DeviceFactory(),
        };
    }

    handleBrands(brandArray) {

        /*
        { "id": 2,
        "name": "创维", 
        "en_name": "Skyworth",
        pinyin":"hai xin"}   
        */
        // console.log('brandList:brandArray:'+ JSON.stringify(brandArray));

        if (brandArray.length == 0) {
            return;
        }
        var allData = [];
        var sectionDict = {}; //包含{"_name":[], cells":{}}

        for (var i = 0; i < brandArray.length; i++) {

            var name = brandArray[i]['name'];//获取拼接 品牌名称
            var pinyin = brandArray[i]['pinyin'];//获取拼接 品牌名称
            var id = brandArray[i]['id'];
            var indexName = MatchTool.getIndexName(pinyin);//索引

            var indexChar = '';
            if (indexName.length > 0) {
                indexChar = indexName.charAt(0);
            }// 太长截取

            if (!sectionDict[indexChar]) {
                sectionDict[indexChar] = [];
            }// 不存在 则添加

            var nameArray = sectionDict[indexChar];//空数组
            var tmpDict = {
                'jsTitle': pinyin,
                'id': id,
                'fullName': name,
            };//字典
            nameArray.push(tmpDict); // 所有品牌数组 

        }

        sectionDict = this.deailNotConatin_Key(sectionDict);
        var keys = this.dealKey_Sort(sectionDict);//获取所有ABCDEFG 的key

        for (var i = 0; i < keys.length; i++) {
            var indexChar = keys[i];
            var nameArray = sectionDict[indexChar];
            nameArray.sort(function(a, b) { return a['name'] == b['name'] });// nuKnow
            var cellDict = [];

            for (var nameIndex = 0; nameIndex < nameArray.length; nameIndex++) {
                /* 自定义属性 customer */
                var customer = (nameArray[nameIndex]['customer'] != undefined) ? (nameArray[nameIndex]['customer']) : (false);

                cellDict.push({
                    "id": nameArray[nameIndex]['id'],
                    "height": 52,
                    "backgroundColor": "#ffffff",
                    "extra": {
                        "separatorColor": "#cccccc",
                        "separatorPadding": 22,
                    },

                    "title": {
                        "text": nameArray[nameIndex]['fullName'],
                        "x": 22,
                        'y': 18,
                        'customer': customer,
                        "fontSize": 15,
                        "foregroundColor": "#4d4d4d",
                    },
                    "jsTitle": nameArray[nameIndex]['jsTitle'],
                });
            }
            allData.push(
                {
                    "header": {
                        "text": keys[i],//sectionArray[sectionIndex],
                        "height": 52,
                        "backgroundColor": "#f8f8f8",
                        "foregroundColor": "#53b99d",
                        'fontIsBold': true,
                        "x": 22,
                        "fontSize": 15,
                        "extra": {
                            "ignoreIndex": false,//sectionIndex == 0,
                            "separatorColor": "#cccccc",
                            "separatorPadding": 22,
                        }
                    },

                    "cells": cellDict,
                },
            );

        }

        var newDataSource = {
            "index": {
                "foregroundColor": "#666666",
            },
            "sections": allData
        };
        this._tableView.reloadData(newDataSource);

    }

    /* 处理不包含 A～ Z字母的 东西*/
    deailNotConatin_Key(sectionDict) {
        var newSectionDict = {};
        var otherArray = [];
        var keys = Object.keys(sectionDict);
        keys.forEach(element => {
            if (!(LetterArray.indexOf(element) == -1)) { /* 在26个字母内*/

                newSectionDict[element] = sectionDict[element];
            } else {/* 不在26个字母内*/


                otherArray = otherArray.concat(sectionDict[element]);

            }
        });

        otherArray.push(this.add_CustomerCell());

        if (otherArray.length > 0) {
            newSectionDict['#'] = otherArray;
        }

        return newSectionDict;
    }

    /* 把 #排到最后一个*/
    dealKey_Sort(sectionDict) {

        var keys = Object.keys(sectionDict);//获取所有ABCDEFG 的key
        keys.sort();//排序
        if (keys.indexOf('#') == 0) {
            var indexChar = keys[0];
            keys.splice(0, 1);

            keys.push(indexChar);

        }
        return keys;
    }

    /* 添加自定义配置按钮 */
    add_CustomerCell() {
        var cellModel = {
            "jsTitle": "自定义",
            "id": undefined,
            "customer": true,
            "fullName": "自定义"
        };
        return cellModel;
    }





    getIPTV_BrandDataSource() {

        var self = this;
        MIRequestServer.Post_GetIPTVBrandList(undefined,
            function(response) {

                if (response.result.brands != undefined) {
                    var dataArray = response.result.brands;
                    sourceData = dataArray;
                    self.handleBrands(dataArray);
                }
            }, function(error) {
                self.refs.toast.showMessage(LocalizedStrings.ir_toast_get_data_failed);
            });
    }


    componentDidMount() {
        /* 获取数据源*/
        this.getIPTV_BrandDataSource();
    }

    render() {
        return (
            /* 列表 */
            <View style={styles.container}>
                <MessageToast ref='toast' />
                <View style={styles.searchView}>
                    <View style={styles.searchContentView}>
                        <View style={styles.searchImageView}>
                            <Image style={styles.searchImage}
                                source={{ isStatic: !MHPluginSDK.devMode, uri: MHPluginSDK.basePath + 'ControllerImage/search.png' }}
                            />
                        </View>
                        <View style={styles.searchInputView}>
                            <TextInput style={styles.searchInput}
                                returnKeyType="search"
                                placeholder={LocalizedStrings.search_device}
                                onFocus={() => this.searchOnFocus()}
                                onChangeText={(value) => this.onChangeText(value)} />
                        </View>
                    </View>
                </View>
                <MHTableView style={styles.table} tableName="brandList" ref={component => this._tableView = component} onSelectRow={this._onSelectRow.bind(this)} />
            </View>

        )
    }

    onChangeText(event) {
        console.log(event);
    }

    searchOnFocus() {

        var self = this;
        this.props.navigator.push({
            ...SearchFilter.route,
            passProps: {
                dataArray: sourceData,
                selectedRow: function(event) {
                    self.pushTo_MatchRemoter(event);
                }
            }
        });

    }

    /*返回首页*/
    backPreviousPage() {
        MHPluginSDK.closeCurrentPage();
    }

    /* 点击 cell 事件*/
    _onSelectRow(event) {
        /* 跳转到对码页面*/
        var self = this;
        var dataSource = this._tableView.getDataSource();
        var obj = dataSource["sections"][event.nativeEvent.selectedSection]["cells"][event.nativeEvent.selectedRow];
        // console.log('brandList-------------obj:'+JSON.stringify(obj));


        /* 0. 如果是自定义的cell 则跳转到自定义学习页面 */
        if (obj.title.customer == true) {
            this.pushTo_CustomerRemoter();
            return;
        }

        /* 1.如果有多个网关跳转到 选择网关页面 */
        this.pushTo_MatchRemoter(obj);

    }


    /* 1.1 跳转到自定义学习页面---适配多个网关*/
    pushTo_CustomerRemoter() {

        var self = this;
        var selectGateWay_ID = this.state.select_GateWayId;
        this.pushCustomerRemoter(selectGateWay_ID);
    }

    /* 1.2 */
    pushCustomerRemoter(selectGateWay_ID) {
        var self = this;
        var deviceType = this.state.deviceType;     /* used */
        var DeviceModel = this.state.device_Factory.GetDeviceModel_WithType(deviceType);

        if (deviceType == 5) {
            this.refs.toast.showMessage('该品类 - 未开通学习功能');
            return;
        }

        self.props.navigator.push({
            ...New_studyRemoter.route,
            passProps: {
                selectGateWay_ID: selectGateWay_ID,
                deviceType: deviceType,
            }
        });

    }

    /* 2.1 push到匹配页面-----适配多个网关 */
    pushTo_MatchRemoter(model) {
        var self = this;
        var brand_id = model.id;
        self.props.selectBrandEvent(brand_id, this.state.select_GateWayId);
    }



};

var screenWidth = Dimensions.get('window').width;
var screenHeight = Dimensions.get('window').height;
const APPBAR_HEIGHT = Platform.OS === 'ios' ? 44 : 56;
const NavigatorBarHeight = (screenHeight == 812) ? (88) : (APPBAR_HEIGHT + 20);
const KEY_OF_MAINPAGE = 'IPTVBrandList';

const searchViewHeight = 50;
var styles = StyleSheet.create({
    centering: {
        alignItems: 'center',
        justifyContent: 'center',
        padding: 8,
    },
    searchInputView: {
        justifyContent: 'center',
        alignItems: 'center',
        alignSelf: 'center',
    },
    searchInput: {
        flex: 1,
        width: screenWidth - 40 - 8,
        height: searchViewHeight - 10,
    },
    searchView: {
        marginTop: NavigatorBarHeight,
        width: screenWidth,
        height: searchViewHeight,
        backgroundColor: 'rgb(238,238,238)',
        justifyContent: 'center',
        alignItems: 'center',
        alignSelf: 'center',
    },
    searchContentView: {
        padding: 5,
        flexDirection: 'row',
        width: screenWidth - 15,
        height: searchViewHeight - 15,
        backgroundColor: '#ffffff',
        justifyContent: 'center',
        alignItems: 'center',
        alignSelf: 'center',
        borderRadius: 5,
    },
    searchImageView: {
        width: 40,
        height: searchViewHeight - 8,
        justifyContent: 'center',
        alignItems: 'center',
        alignSelf: 'center',
    },
    searchImage: {
        width: 18,
        height: 18,
    },
    NoneGateWayContainer: {
        flex: 1,
        backgroundColor: '#ffffff',
        marginBottom: 0,
        marginTop: 0,
        flexDirection: 'column',
        justifyContent: 'center',
        alignItems: 'center',
        alignSelf: 'center',
    },
    container: {
        flex: 1,
        // flexDirection: 'column',
        // justifyContent: 'center',
        // alignItems: 'center',
        backgroundColor: '#f8f8f8',
        marginBottom: 0,
        marginTop: 0,
    },
    containerView: {
        flexDirection: 'column',
        justifyContent: 'center',
        alignItems: 'center',
    },
    titleMessage: {
        textAlign: 'center',
        fontSize: 14,
        color: '#808080',
    },
    jumpBack: {
        textAlign: 'center',
        fontSize: 14,
        color: 'rgb(52,117,232)',
    },
    section: {
        height: 20,
        fontSize: 15,
        flexDirection: 'row',
        borderWidth: 1,
        borderColor: '#efeff0',
        borderBottomColor: '#AAA',
        color: '#979799',
        backgroundColor: '#efeff0',
    },
    sectionText: {
        marginTop: 2,
        marginLeft: 20,
        // width:200,
    },
    row: {
        marginLeft: 22,
        marginTop: 20,
        height: 44,
        fontSize: 18,
        color: '#555'

    },
    border: {
        height: 1,
        backgroundColor: '#aaaaaa'
    },
    text: {
        flex: 1,
        fontSize: 20,
        textAlign: 'center',
        color: '#000000',
        alignSelf: 'stretch',
        marginTop: 300,
    },
    table: {
        marginTop: 0,
        flex: 1,
        backgroundColor: '#f8f8f8',
    },
});


var route = {
    key: KEY_OF_MAINPAGE,
    title: LocalizedStrings.brand_list,
    component: IPTVBrandList,
    navLeftButtonStyle: {
        tintColor: 'rgb(236,236,236)',
    },
    navTitleStyle: {
        color: '#808080',
    },
    navBarStyle: {
        backgroundColor: '#f8f8f8',
    },
    isNavigationBarHidden: false,
    renderNavLeftComponent: function(route, navigator, index, navState) {
        return (
            <View style={{ left: 0, width: 29 + 15 * 2, height: APPBAR_HEIGHT, justifyContent: 'center', alignItems: 'center' }}>
                <ImageButton
                    source={{ uri: MHPluginSDK.uriNaviBackButtonImage, scale: PixelRatio.get() }}
                    onPress={() => {
                        if (index === 0) {
                            MHPluginSDK.closeCurrentPage();
                        } else {
                            navigator.pop();
                        }
                    }}
                    style={[{ width: 29, height: 29, tintColor: 'rgb(0,0,0)' }]}
                />
            </View>
        );
    },
    renderNavRightComponent: function(route, navigator, index, navState) {
        return null;
    },
}


module.exports = {
    component: IPTVBrandList,
    route: route
};
