// mh-infrared-gateway/MainV2/AllRemoter/DVDBoxRemote.js
'use strict';

var React = require('react-native');

var {
  AppRegistry,
  StyleSheet,
  Text,
  TouchableHighlight,
  Image,
  View,
  TextInput,
  PixelRatio,
  StatusBar,
  TouchableOpacity,
  Platform,
  ScrollView,
  Dimensions,
  DeviceEventEmitter,
} = React;

import { NativeModules } from 'react-native';
var MHPluginSDK = NativeModules.MHPluginSDK;
var Setting = require('../../Main/MHSetting');
var ImageButton = require('../../CommonModules/ImageButton');
var Button = require('../../CommonModules/Button');
var MIRequestServer = require('../../MainV2/MIRequestServer/MIRequestServer').MIRequestServer;
var LocalizedStrings = require('../../Main/MHLocalizableString.js').string;
var DirectionView = require('../../MainV2/ToolModel/DirectionView');
var NormalExtendView = require('../../MainV2/ToolModel/NormalExtendView');
var NavigatorView = require('../../MainV2/ToolModel/NavigatorView');
var MessageToast = require('../../MainV2/ToolModel/MessageToast');

var screenWidth = Dimensions.get('window').width;
var screenHeight = Dimensions.get('window').height;
const APPBAR_HEIGHT = Platform.OS === 'ios' ? 44 : 56;
const NavigatorBarHeight = (screenHeight==812)?(88):( APPBAR_HEIGHT +20);
const SafeBottomHeight =  (screenHeight==812)?(17):(0);

class DVDBoxVC extends React.Component {
  constructor(props, context) {
    super(props, context);
    var deviceId =  (props.did!=undefined)?(props.did):(MHPluginSDK.deviceId);
    
    this.state={
      did: deviceId,
      model: MHPluginSDK.deviceModel,
      apiLevel: MHPluginSDK.apiLevel,
      basePath: MHPluginSDK.basePath,
      devMode: MHPluginSDK.devMode,
      remoterID:this.props.remoterID,
      isShow:false,
      allKeys:[],
    };
    this.enableExtendButton =false;
    console.log('--------DVDBoxVC-this.props-------------'+JSON.stringify(this.state));
   
  }


  render(){

    return (
      <View style={styles.containerAll}>
          {/* 第一行 */}
          <MessageToast ref='toast'/>
          <Image style={[styles.backgroundImage,{'position':'absolute',width:screenWidth,height:screenHeight}]}
              source={{isStatic:!MHPluginSDK.devMode, uri:MHPluginSDK.basePath + 'NewControllerImage/normal_bg.png'}}
              /> 
          <NormalExtendView isShow={this.state.isShow} 
          onPressButton={(event)=>this.onPressButton(event)} 
          onPressDissmissButton={(event)=>this.onPressDissmissButton(event)}
          dataSource ={this.deal_ExtendViewData(this.state.allKeys)} 
          />    
          <NavigatorView ref='navigator' title={route.title} />  
          <View style={styles.scrollContentView}>
                <View style={styles.TopMenuView}>
                      <View style={styles.middleButtonView}>
                          <View style={styles.threeButtonView}>
                                <Button  onPress={()=>{this.onPressButton('POWER')}} enabled={this.isButtonEnabled('POWER')} shock ={true} imageNormal="NewControllerImage/tv_power_normal.png" imageHighlight='NewControllerImage/tv_power_press.png' imageWidth={60} imageHeight={60} />
                          </View> 
                          <View style={styles.threeButtonView}>
                                <Button   onPress={()=>{this.onPressButton('STOP')}} enabled={this.isButtonEnabled('STOP')} shock ={true}  imageNormal="NewControllerImage/dvd_stop_normal.png" imageHighlight='NewControllerImage/dvd_stop_press.png' imageWidth={60} imageHeight={60} />
                          </View> 
                          <View style={styles.threeButtonView}>
                                <Button   onPress={()=>{this.onPressButton('PLAY')}} enabled={this.isButtonEnabled('PLAY')} shock ={true} imageNormal="NewControllerImage/tv_play_normal.png" imageHighlight='NewControllerImage/tv_play_press.png' imageWidth={60} imageHeight={60} />
                          </View>   
                      </View>
                      <View style={styles.middleButtonView}>
                          <View style={styles.fourButtonView}>
                                <Button  onPress={()=>{this.onPressButton('PREVIOUS')}} enabled={this.isButtonEnabled('PREVIOUS')} shock ={true} imageNormal="NewControllerImage/dvd_PREVIOUS_normal.png" imageHighlight='NewControllerImage/dvd_PREVIOUS_press.png' imageWidth={60} imageHeight={60} />
                          </View> 
                          <View style={styles.fourButtonView}>
                                <Button   onPress={()=>{this.onPressButton('REWIND')}} enabled={this.isButtonEnabled('REWIND')} shock ={true}  imageNormal="NewControllerImage/tv_rewind_normal.png" imageHighlight='NewControllerImage/tv_rewind_press.png' imageWidth={60} imageHeight={60} />
                          </View> 
                          <View style={styles.fourButtonView}>
                                <Button   onPress={()=>{this.onPressButton('FAST_FORWARD')}} enabled={this.isButtonEnabled('FAST_FORWARD')} shock ={true} imageNormal="NewControllerImage/tv_fast_forward_normal.png" imageHighlight='NewControllerImage/tv_fast_forward_press.png' imageWidth={60} imageHeight={60} />
                          </View>
                          <View style={styles.fourButtonView}>
                                <Button   onPress={()=>{this.onPressButton('NEXT')}} enabled={this.isButtonEnabled('NEXT')} shock ={true} imageNormal="NewControllerImage/dvd_NEXT_normal.png" imageHighlight='NewControllerImage/dvd_NEXT_press.png' imageWidth={60} imageHeight={60} />
                          </View>      
                      </View>
                </View>
                {/* 2.方向盘 */}
                <View style={styles.middleDirectionView}>
                    <View style={styles.directionContentView}>
                        <DirectionView style={[styles.directionContentView,{'width':210,'height':210}]} 
                        width={210} height={210}
                        dataSource={this.state.allKeys}
                        onPressButton={(event)=>{this.onPressButton(event)}}
                        />
                    </View>
                </View>
                {/* 3.底部 */}
                <View style={styles.bottomView}>
                    <View style={styles.bottomLeftView}>
                        <Button   onPress={()=>{this.onPressButton('VOL+')}} enabled={this.isButtonEnabled('VOL+')} shock ={true} imageNormal="NewControllerImage/dvd_vol+_normal.png" imageHighlight='NewControllerImage/dvd_vol+_press.png' imageWidth={80} imageHeight={82} />
                        <Button   onPress={()=>{this.onPressButton('VOL-')}} enabled={this.isButtonEnabled('VOL-')} shock ={true} imageNormal="NewControllerImage/dvd_vol-_normal.png" imageHighlight='NewControllerImage/dvd_vol-_press.png' imageWidth={80} imageHeight={80} />
                    </View>
                    <View style={styles.bottomRightView}>
                        <Button   onPress={()=>{this.onPressExtendButton('EXTENSION')}} enabled={this.isEnableExtendButton()}  shock ={true} imageNormal="NewControllerImage/tv_extend_normal_special.png" imageHighlight='NewControllerImage/camera_extend_press.png' imageWidth={60} imageHeight={60} />
                    </View>
                </View>               
          </View>   
      </View>
    );
  }

  onPressButton(event) {

        this.refs.navigator.show();
        console.log('onPressButton------outSide------'+event);
        var key_id = undefined;
        this.state.allKeys.forEach(element => {
          
            if(element.name==event){
              key_id = element.id;
            }
        });
        console.log('key_id:'+key_id +'\n'+'event:'+event);

        if(key_id!=undefined){
        
          var self = this;
          var param  = {
            "did":this.state.did,
            controller_id:this.state.remoterID,
            "key_id": key_id
          };

          MIRequestServer.Post_SendkeyCode(param,
            function(response){
                if((response.code !=undefined) &&(response.code==0)){
                  // self.refs.toast.showMessage(response.message);
                }
                
            },function(error){
                console.log('Post_SendkeyCode-error---'+JSON.stringify(error));
            });  
        }

  }

  isButtonEnabled(event) {

      var key_id = undefined;
      this.state.allKeys.forEach(element => {
          if(element.name==event){
            key_id = element.id;
          }
      });

      if(key_id!=undefined){
        return true;
      }else{
        return false;
      }
    
  }

  componentWillMount() {
     /*后期改成 controller_id */
     var param  = {"did": this.state.did};
     var self = this;
 
     MIRequestServer.Post_GetAllRemoterkey(param,
       function(response){
          
           if((response.result.keys)&&(response.result.keys!=undefined)){
             var allKeys = response.result.keys;
             self.setState({ allKeys:allKeys});
           }
 
       },function(error){
          self.refs.toast.showMessage(LocalizedStrings.ir_toast_get_data_failed);
       });  
  }

  componentWillUnmount() {

  }

  deal_ExtendViewData(allKeys){

      var bindKeyArray =['POWER','STOP','PLAY','PREVIOUS',
      'REWIND','FAST_FORWARD','NEXT','VOL+','VOL-',
      'NAVIGATE_UP','NAVIGATE_DOWN','NAVIGATE_LEFT','NAVIGATE_RIGHT','OK'];
      var NewAllKeys = allKeys.slice(0);
      var deleteArray =[];
      NewAllKeys.forEach((element,index) => {
          var isEqual = false;
          var sourceName = element.name;

          for(let i=0;i<bindKeyArray.length;i++){
              var key = bindKeyArray[i];
              if(key ==sourceName ){
                isEqual = true; 
                deleteArray.push(element);

              }
          }
      }); 
      deleteArray.forEach((element) => {

        let index = NewAllKeys.indexOf(element);      
        NewAllKeys.splice(index,1);
      });
      if(NewAllKeys.length>0){
        this.enableExtendButton =true;
      }
      return NewAllKeys;  
  }

  isEnableExtendButton(){
    return this.enableExtendButton;
  }
  
  onPressExtendButton(){
    this.setState({isShow:true});    
  }

  onPressDissmissButton(){
    this.setState({isShow:false});
  }          
}

const sectionHeght =80;
const sectionContentWidth=  screenWidth- 20*2;
const ButtonViewWidth=  screenWidth/3.0;
const imageViewWidth = 240;
const imageViewHeight = imageViewWidth;
const bottomViewHeight = 120;
const directionHeight =(screenHeight-NavigatorBarHeight-SafeBottomHeight) -(bottomViewHeight) -(sectionHeght*2);


var styles = StyleSheet.create({
  containerAll: {
    flex:1,
    alignItems:'center',
    alignSelf:'center',
  },
  backgroundImage:{
    marginTop:0,
    width:screenWidth,
    height:screenHeight,
  },
  extendModalView:{
    flex:1,
    backgroundColor:'#ffffff',
    alignItems:'center',
    justifyContent:'center'
  },
  TopMenuView:{
    height:sectionHeght*2,
    width:screenWidth,
  },
  middleDirectionView:{
    width:screenWidth,
    height:directionHeight,
    alignItems:'center',
    alignSelf:'center',
    justifyContent:'center',
  },
  bottomView:{
    height:bottomViewHeight,
    width:sectionContentWidth,
    alignItems:'center',
    alignSelf:'center',
    flexDirection:'row',
  },
  bottomLeftView:{
    width:(sectionContentWidth*2)/3,
    height:bottomViewHeight,
    justifyContent:'center',
    alignItems:'center',
    alignSelf:'center',
    flexDirection:'row',
  },
  bottomRightView:{
    width:(sectionContentWidth)/3,
    height:bottomViewHeight,
    justifyContent:'center',
    alignItems:'center',
    alignSelf:'center',
  },
  scrollContentView:{
    marginTop:0,
    width:screenWidth,
    height:screenHeight-NavigatorBarHeight-SafeBottomHeight,
    marginBottom:SafeBottomHeight,
  },
  directionContentView:{
    width:imageViewWidth,
    height:imageViewHeight,
    flexDirection:'column',
    justifyContent:'center',
    alignItems:'center',
    alignSelf:'center',
  },
  imageView:{
    width:imageViewWidth,
    height:imageViewWidth,
  },
  middleButtonView:{
    alignItems:'center',
    alignSelf:'center',
    flexDirection:'row',
    width:sectionContentWidth,
    height:sectionHeght,
    marginBottom:3,
  },
  threeButtonView:{
    flexDirection:'row',
    width:sectionContentWidth/3,
    height:sectionHeght,
    alignItems:'center',
    alignSelf:'center',
    justifyContent:'center',
  },
  fourButtonView:{
    flexDirection:'row',
    width:sectionContentWidth/4,
    height:sectionHeght,
    alignItems:'center',
    alignSelf:'center',
    justifyContent:'center',
  }
});

const KEY_OF_MAINPAGE = 'DVDBoxVC';

// 打开更多菜单
var openMorePage = function (navigator) {
  navigator.push(Setting.route);
};

// 每个页面export自己的route
var route = {
  key: KEY_OF_MAINPAGE,
  title: MHPluginSDK.deviceName,
  component: DVDBoxVC,
  navLeftButtonStyle: {
    tintColor:'#ffffff',
  },
  navTitleStyle: {
    color:'#ffffff',
  },
  navBarStyle: {
    backgroundColor:'transparent',
  },
  isNavigationBarHidden: false,
  renderNavLeftComponent: function(route, navigator, index, navState) {
    return (
      <View style={{left:0, width:29+15*2, height:APPBAR_HEIGHT, justifyContent:'center', alignItems:'center'}}>
        <ImageButton
          source={{uri:MHPluginSDK.uriNaviBackButtonImage, scale:PixelRatio.get()}}
          onPress={() => {
                if (index === 0) {
                  MHPluginSDK.closeCurrentPage();
                } else {
                  navigator.pop();
                }
          }}
          style={[{width:29, height:29, tintColor: '#ffffff'}]}
        />
      </View>
    );
  },
  renderNavRightComponent: function(route, navigator, index, navState) {
    if ((MHPluginSDK.userId == MHPluginSDK.ownerId)&&(index==0)) // 非分享设备
    {
      return (
        <View style={{left:0, width:29+15*2, height:APPBAR_HEIGHT, justifyContent:'center', alignItems:'center'}}>
          <ImageButton
            source={{uri:MHPluginSDK.uriNaviMoreButtonImage, scale:PixelRatio.get()}}
            onPress={() => {
              openMorePage(navigator);
            }}
            style={[{width:29, height:29, tintColor: '#ffffff'}]}
          />
        </View>
      );
    }
    else {
      return null;
    }
  },
}

module.exports = {
  component: DVDBoxVC,
  route: route,
}
