// mh-infrared-gateway/MainV2/AllRemoter/ElectricFanBoxRemote.js
'use strict';

var React = require('react-native');

var {
  AppRegistry,
  StyleSheet,
  Text,
  TouchableHighlight,
  Image,
  View,
  Switch,
  TextInput,
  PixelRatio,
  StatusBar,
  TouchableOpacity,
  Platform,
  Dimensions,
  ScrollView,
  Button,
  DeviceEventEmitter,
} = React;



import { NativeModules } from 'react-native';
var MHPluginSDK = NativeModules.MHPluginSDK;
var Setting = require('../../Main/MHSetting');
var ImageButton = require('../../CommonModules/ImageButton');
var Button = require('../../CommonModules/Button');
var TitleButton = require('../../CommonModules/TitleButton');
var MIRequestServer = require('../../MainV2/MIRequestServer/MIRequestServer').MIRequestServer;
var NormalExtendView = require('../../MainV2/ToolModel/NormalExtendView');
var MessageToast = require('../../MainV2/ToolModel/MessageToast');
var NavigatorView = require('../../MainV2/ToolModel/NavigatorView');
var LocalizedStrings = require('../../Main/MHLocalizableString.js').string;

var screenWidth = Dimensions.get('window').width;
var screenHeight = Dimensions.get('window').height;
const APPBAR_HEIGHT = Platform.OS === 'ios' ? 44 : 56;
const NavigatorBarHeight = (screenHeight==812)?(88):( APPBAR_HEIGHT +20);
const SafeBottomHeight =  (screenHeight==812)?(17):(0);


var modes = ['mode_change_air.png','mode_wind.png','mode_dry.png','mode_heat.png','mode_cool.png'];


class EletricFanBoxVC extends React.Component {
  constructor(props, context) {
    super(props, context);
    var deviceId =  (props.did!=undefined)?(props.did):(MHPluginSDK.deviceId);
    
    this.state = {
            did: deviceId,
            model: MHPluginSDK.deviceModel,
            apiLevel: MHPluginSDK.apiLevel,
            basePath: MHPluginSDK.basePath,
            devMode: MHPluginSDK.devMode,
            isShow:false,
            isOnline:false,
            allKeys:[],
        };
    var self = this;
    this.enableExtendButton =false;
    MHPluginSDK.getDevicePropertyFromMemCache(['isOnline'],(response)=>{

      self.isOnline== response.isOnline;
      self.setState({isOnline:response.isOnline});
    });
    console.log('--------EletricFanBoxVC-this.props-------------'+JSON.stringify(this.state));
  }

  render(){
    return (
      <View style={styles.containerAll}>
          <MessageToast ref='toast'/>
          <NormalExtendView isShow={this.state.isShow} 
            onPressButton={(event)=>this.onPressButton(event)} 
            onPressDissmissButton={(event)=>this.onPressDissmissButton(event)}
            dataSource ={this.deal_ExtendViewData(this.state.allKeys)} 
          />   
          {
            // this.offLine_statusView()
            (this.state.isOnline==true)?(this.onLine_statusView()):(this.offLine_statusView())
          }
         <NavigatorView ref='navigator' title={route.title} />
          <View style={styles.contentBgView}>
              <View style={styles.contentBottomView}>
                  <View style={styles.middleButtonView}>
                      <View style={styles.fiveButtonView}>
                            <Button onPress={()=>{this.onPressButton('TIMER')}}  imageNormal="NewControllerImage/eletricfan_timer_normal.png" imageHighlight="NewControllerImage/eletricfan_timer_press.png" enabled={this.isButtonEnabled('TIMER')} style={{marginBottom:8}} title ={LocalizedStrings.timer} shock ={true} textColor={'#ffffff'} titleSize={15} imageWidth={45} imageHeight={45} />
                      </View> 
                      <View style={styles.fiveButtonView}>
                            <Button onPress={()=>{this.onPressButton('SWING')}} imageNormal="NewControllerImage/eletricfan_swing_normal.png" imageHighlight="NewControllerImage/eletricfan_swing_press.png" enabled={this.isButtonEnabled('SWING')} style={{marginBottom:8}} title ={LocalizedStrings.wind_sweep}  shock ={true} textColor={'#ffffff'} titleSize={15}  imageWidth={45} imageHeight={45}/>
                      </View>  
                      <View style={styles.fiveButtonView}>
                            <Button onPress={()=>{this.onPressButton('FAN_SPEED')}} imageNormal="NewControllerImage/airpuriffer_speed_normal.png" imageHighlight="NewControllerImage/airpuriffer_speed_press.png" enabled={this.isButtonEnabled('FAN_SPEED')} style={{marginBottom:8}} title ={LocalizedStrings.wind_speed}  shock ={true} textColor={'#ffffff'} titleSize={15}  imageWidth={45} imageHeight={45}/>
                      </View>  
                      <View style={styles.fiveButtonView}>
                            <Button onPress={()=>{this.onPressButton('SWING_MODE')}} imageNormal="NewControllerImage/eletricfan_speed_direction_normal.png" imageHighlight="NewControllerImage/eletricfan_speed_direction_press.png" enabled={this.isButtonEnabled('SWING_MODE')} style={{marginBottom:8}} title ={LocalizedStrings.ir_controller_wind_class_text}  shock ={true} textColor={'#ffffff'} titleSize={15} imageWidth={45} imageHeight={45} />
                      </View> 
                      <View style={styles.fiveButtonView}>
                            <Button onPress={()=>{this.onPressExtendButton('EXTENSION')}} enabled={this.isEnableExtendButton()} enabled={this.isEnableExtendButton()} imageNormal="NewControllerImage/airpuriffer_extend_normal.png" imageHighlight="NewControllerImage/airpuriffer_extend_press.png"  style={{marginBottom:8}} title ={LocalizedStrings.extra_key_text}  shock ={true} textColor={'#ffffff'} titleSize={15} imageWidth={45} imageHeight={45}  />
                      </View> 
                  </View>
                  <View style={[styles.middleButtonView,{height:100}]}>
                      <View style={styles.threeButtonView}>
                      </View> 
                      <View style={styles.threeButtonView}>
                          <Button onPress={()=>{this.onPressButton('POWER')}} imageNormal="NewControllerImage/airpuriffer_power_normal.png" imageHighlight="NewControllerImage/airpuriffer_power_press.png"  enabled={this.isButtonEnabled('POWER')}  shock ={true}   imageWidth={80} imageHeight={80} />
                      </View> 
                      <View style={styles.threeButtonView}>
                      </View> 
                      
                  </View>
              </View>              
          </View>

      </View>
    );
  }

  offLine_statusView(){
      return(
      <View>
        <Image style={[styles.backgroundImage,{'position':'absolute',width:screenWidth,height:screenHeight}]}
               source={{isStatic:!MHPluginSDK.devMode, uri:MHPluginSDK.basePath + 'NewControllerImage/offline_bg.png'}}/> 
        <Image style={[styles.backgroundImage,{'position':'absolute',width:screenWidth,height:screenHeight}]}
               source={{isStatic:!MHPluginSDK.devMode, uri:MHPluginSDK.basePath + 'NewControllerImage/eletricfan_box_offline.png'}}/>
      </View>);
  }

  onLine_statusView(){
    return(
      <View>
        <Image style={[styles.backgroundImage,{'position':'absolute',width:screenWidth,height:screenHeight}]}
               source={{isStatic:!MHPluginSDK.devMode, uri:MHPluginSDK.basePath + 'NewControllerImage/eletricfan_bg.png'}}/> 
        <Image style={[styles.backgroundImage,{'position':'absolute',width:screenWidth,height:screenHeight}]}
               source={{isStatic:!MHPluginSDK.devMode, uri:MHPluginSDK.basePath + 'NewControllerImage/eletricfan_box_offline.png'}}/>
      </View>);
  }

  onPressButton(event) {

      this.refs.navigator.show();
      var key_id = undefined;
      this.state.allKeys.forEach(element => {
        
          if(element.name==event){
            key_id = element.id;
          }
      });
      console.log('key_id:'+key_id +'\n'+'event:'+event);

      if(key_id!=undefined){
      
        var self = this;
        var param  = {
          "did":this.state.did,
          controller_id:this.state.remoterID,
          "key_id": key_id
        };

        MIRequestServer.Post_SendkeyCode(param,
          function(response){
            if((response.code !=undefined) &&(response.code==0)){
              // self.refs.toast.showMessage(response.message);
            }
              
          },function(error){
            console.log('Post_SendkeyCode-error---'+JSON.stringify(error));
          });  
      }


  }

  isButtonEnabled(event) {

        var key_id = undefined;
        this.state.allKeys.forEach(element => {
            if(element.name==event){
              key_id = element.id;
            }
        });

        if(key_id!=undefined){
          return true;
        }else{
          return false;
        }
    
  }

  componentWillMount() {
     /*后期改成 controller_id */
     var param  = {"did": this.state.did};
     var self = this;
 
     MIRequestServer.Post_GetAllRemoterkey(param,
       function(response){
          
           if((response.result.keys)&&(response.result.keys!=undefined)){
             var allKeys = response.result.keys;
             self.setState({ allKeys:allKeys});
           }
 
       },function(error){
          self.refs.toast.showMessage(LocalizedStrings.ir_toast_get_data_failed);
       });  
  }

  componentWillUnmount() {

  }

  deal_ExtendViewData(allKeys){

      var bindKeyArray =['POWER','TIMER','SWING','FAN_SPEED','SWING_MODE',];
      let NewAllKeys = allKeys.slice(0);
      var deleteArray =[];


      NewAllKeys.forEach((element,index) => {
          var isEqual = false;
          var sourceName = element.name;
          
          for(let i=0;i<bindKeyArray.length;i++){
              var key = bindKeyArray[i];
              
              if(sourceName ==key ){
                isEqual = true; 
                deleteArray.push(element);
              }
          }
      }); 
      
      /*删除 循环 */
      deleteArray.forEach((element) => {

        let index = NewAllKeys.indexOf(element);      
        NewAllKeys.splice(index,1);
      });

      if(NewAllKeys.length>0){
        this.enableExtendButton =true;
      }
      return NewAllKeys;  
  }

  isEnableExtendButton(){
    return this.enableExtendButton;
  }

  onPressExtendButton(){
    this.setState({isShow:true});    
  }

  onPressDissmissButton(){
    this.setState({isShow:false});
  } 
}

const sectionHeght =( screenHeight - NavigatorBarHeight -300)/3.0;
const sectionWidth=  screenWidth/2.0;
const ThreeButtonViewWidth=  screenWidth/3.0;
const FourButtonViewWidth =  screenWidth/4.0;
const FiveButtonViewWidth =  screenWidth/5.0;
const contentBottomViewMarginTop = (screenHeight-NavigatorBarHeight-SafeBottomHeight)-(100+sectionHeght);

var styles = StyleSheet.create({
  containerAll: {
    flex:1,
  },
  backgroundImage:{
    width:screenWidth,
    height:screenHeight,
  },
  contentBgView:{
    marginTop:0,
    height:screenHeight-NavigatorBarHeight,
    width:screenWidth,
  },
  contentBottomView:{
    marginBottom:50,
    marginTop:contentBottomViewMarginTop,
  },
  threeButtonView:{
    flexDirection:'row',
    width:ThreeButtonViewWidth,
    height:sectionHeght,
    alignItems:'center',
    alignSelf:'center',
    justifyContent:'center',
  },
  fourButtonView:{
    flexDirection:'row',
    width:FourButtonViewWidth,
    height:sectionHeght,
    alignItems:'center',
    alignSelf:'center',
    justifyContent:'center',
  },
  fiveButtonView:{
    flexDirection:'row',
    width:FiveButtonViewWidth,
    height:sectionHeght,
    alignItems:'center',
    alignSelf:'center',
    justifyContent:'center',
  },
  middleButtonView:{
    flexDirection:'row',
    width:screenWidth,
    height:sectionHeght,
    alignItems:'center',
    justifyContent:'center',
  },
  topImage:{
    marginTop:NavigatorBarHeight,
    width:screenHeight,
    height:300,
    justifyContent:'center',
    alignItems:'center',
    alignSelf:'center',
    backgroundColor:'rgb(238,238,238)',
  },
  image: {
    alignSelf: 'center',
    resizeMode: 'contain',
  },
  menuView:{
    width:screenWidth,
    height:sectionHeght,
    flexDirection:'row',
  },
  

});

const KEY_OF_MAINPAGE = 'EletricFanBoxVC';

// 打开更多菜单
var openMorePage = function (navigator) {
  navigator.push(Setting.route);
};

// 每个页面export自己的route
var route = {
  key: KEY_OF_MAINPAGE,
  title: MHPluginSDK.deviceName,
  component: EletricFanBoxVC,
  navLeftButtonStyle: {
    tintColor:'#ffffff',
  },
  navTitleStyle: {
    color:'#ffffff',
  },
  navBarStyle: {
    backgroundColor:'transparent',
  },
  isNavigationBarHidden: false,
  renderNavLeftComponent: function(route, navigator, index, navState) {
    return (
      <View style={{left:0, width:29+15*2, height:APPBAR_HEIGHT, justifyContent:'center', alignItems:'center'}}>
        <ImageButton
          source={{uri:MHPluginSDK.uriNaviBackButtonImage, scale:PixelRatio.get()}}
          onPress={() => {
                if (index === 0) {
                  MHPluginSDK.closeCurrentPage();
                } else {
                  navigator.pop();
                }
          }}
          style={[{width:29, height:29, tintColor: '#ffffff'}]}
        />
      </View>
    );
  },
  renderNavRightComponent: function(route, navigator, index, navState) {
    if ((MHPluginSDK.userId == MHPluginSDK.ownerId)&&(index==0)) // 非分享设备
    {
      return (
        <View style={{left:0, width:29+15*2, height:APPBAR_HEIGHT, justifyContent:'center', alignItems:'center'}}>
          <ImageButton
            source={{uri:MHPluginSDK.uriNaviMoreButtonImage, scale:PixelRatio.get()}}
            onPress={() => {
              openMorePage(navigator);
            }}
            style={[{width:29, height:29, tintColor: '#ffffff'}]}
          />
        </View>
      );
    }
    else {
      return null;
    }
  },
}

module.exports = {
  component: EletricFanBoxVC,
  route: route,
}
