// mh-infrared-gateway/MainV2/AllRemoter/HeaterBoxRemoter.js
'use strict';

var React = require('react-native');

var {
  AppRegistry,
  StyleSheet,
  Text,
  TouchableHighlight,
  Image,
  View,
  TextInput,
  PixelRatio,
  StatusBar,
  Dimensions,
  TouchableOpacity,
  Platform,
  Switch,
  DeviceEventEmitter,
} = React;

import { NativeModules } from 'react-native';
var MHPluginSDK = NativeModules.MHPluginSDK;
var Setting = require('../../Main/MHSetting');
var Button = require('../../CommonModules/Button');
var ImageButton = require('../../CommonModules/ImageButton');
var TitleButton = require('../../CommonModules/TitleButton');
var MessageToast = require('../../MainV2/ToolModel/MessageToast');
var MIRequestServer = require('../../MainV2/MIRequestServer/MIRequestServer').MIRequestServer;
var NormalExtendView = require('../../MainV2/ToolModel/NormalExtendView');
var NavigatorView = require('../../MainV2/ToolModel/NavigatorView');
var LocalizedStrings = require('../../Main/MHLocalizableString.js').string;

var screenWidth = Dimensions.get('window').width;
var screenHeight = Dimensions.get('window').height;
const APPBAR_HEIGHT = Platform.OS === 'ios' ? 44 : 56;
const NavigatorBarHeight = (screenHeight==812)?(88):( APPBAR_HEIGHT +20);
const SafeBottomHeight =  (screenHeight==812)?(17):(0);

class HeaterBoxVC extends React.Component {
  constructor(props, context) {
    super(props, context);
    var deviceId =  (props.did!=undefined)?(props.did):(MHPluginSDK.deviceId);
    
    this.state={
      did: deviceId,
      model: MHPluginSDK.deviceModel,
      apiLevel: MHPluginSDK.apiLevel,
      basePath: MHPluginSDK.basePath,
      devMode: MHPluginSDK.devMode,
      remoterID:this.props.remoterID,
      isShow:false,
      isOnline:false,
      allKeys:[],
    };
    this.enableExtendButton =false;
    var self = this;
    MHPluginSDK.getDevicePropertyFromMemCache(['isOnline'],(response)=>{

      self.isOnline== response.isOnline;
      self.setState({isOnline:response.isOnline});
    });
    console.log('--------HeaterBoxVC-this.props-------------'+JSON.stringify(this.state));
  }

  render(){
    return (
      <View style={styles.containerAll}>
          <MessageToast ref='toast'/>
          <NormalExtendView isShow={this.state.isShow} 
          onPressButton={(event)=>this.onPressButton(event)} 
          onPressDissmissButton={(event)=>this.onPressDissmissButton(event)}
          dataSource ={this.deal_ExtendViewData(this.state.allKeys)} 
          />
          {
            // this.offLine_statusView()
            (this.state.isOnline==true)?(this.onLine_statusView()):(this.offLine_statusView())
          }
           <NavigatorView ref='navigator' title={route.title} />
           
          <View style={styles.contentBgView}>
              <View style={styles.contentBottomView}>
                  <View style={styles.middleButtonView}>
                      <View style={styles.fourButtonView}>
                            <Button onPress={()=>{this.onPressButton('AUTO')}}  imageNormal="NewControllerImage/heater_mode_normal.png" imageHighlight="NewControllerImage/heater_mode_press.png" enabled={this.isButtonEnabled('AUTO')} style={{marginBottom:8}} title ={LocalizedStrings.ir_controller_mode_text} shock ={true} textColor={'#ffffff'} titleSize={15} imageWidth={45} imageHeight={45}  />
                      </View> 
                      <View style={styles.fourButtonView}>
                            <Button onPress={()=>{this.onPressButton('TEMP+')}} imageNormal="NewControllerImage/heater_tep+_normal.png" imageHighlight="NewControllerImage/heater_tep+_press.png" enabled={this.isButtonEnabled('TEMP+')} style={{marginBottom:8}} title ={LocalizedStrings.ir_controller_temp_up_text}  shock ={true} textColor={'#ffffff'} titleSize={15} imageWidth={45} imageHeight={45} />
                      </View>  
                      <View style={styles.fourButtonView}>
                            <Button onPress={()=>{this.onPressButton('TEMP-')}} imageNormal="NewControllerImage/heater_tep-_normal.png" imageHighlight="NewControllerImage/heater_tep-_press.png" enabled={this.isButtonEnabled('TEMP-')} style={{marginBottom:8}} title ={LocalizedStrings.ir_controller_temp_down_text}  shock ={true} textColor={'#ffffff'} titleSize={15} imageWidth={45} imageHeight={45} />
                      </View>  
                      <View style={styles.fourButtonView}>
                            <Button onPress={()=>{this.onPressExtendButton('EXTENSION')}}  enabled={this.isEnableExtendButton()} imageNormal="NewControllerImage/airpuriffer_extend_normal.png" imageHighlight="NewControllerImage/airpuriffer_extend_press.png"  style={{marginBottom:8}} title ={LocalizedStrings.extra_key_text}  shock ={true} textColor={'#ffffff'} titleSize={15} imageWidth={45} imageHeight={45} />
                      </View> 
                  </View>
                  <View style={[styles.middleButtonView,{height:100}]}>
                      <View style={styles.threeButtonView}>
                      </View> 
                      <View style={styles.threeButtonView}>
                          <Button onPress={()=>{this.onPressButton('POWER')}} imageNormal="NewControllerImage/airpuriffer_power_normal.png" imageHighlight="NewControllerImage/airpuriffer_power_press.png"  enabled={this.isButtonEnabled('POWER')}  shock ={true}   imageWidth={80} imageHeight={80} />
                      </View> 
                      <View style={styles.threeButtonView}>
                      </View> 
                  </View>
              </View>              
          </View>
      </View>
    );
  }

  onLine_statusView(){

      return(
        <View>
          <Image style={[styles.backgroundImage,{'position':'absolute',width:screenWidth,height:screenHeight}]}
                source={{isStatic:!MHPluginSDK.devMode, uri:MHPluginSDK.basePath + 'NewControllerImage/heater_bg_online.png'}} /> 
            <Image style={[styles.backgroundImage,{'position':'absolute',width:screenWidth,height:screenHeight}]}
                source={{isStatic:!MHPluginSDK.devMode, uri:MHPluginSDK.basePath + 'NewControllerImage/heater_box_online.png'}}/> 
        </View>
      );
  }

  offLine_statusView(){

      return(
        <View>
          <Image style={[styles.backgroundImage,{'position':'absolute',width:screenWidth,height:screenHeight}]}
                source={{isStatic:!MHPluginSDK.devMode, uri:MHPluginSDK.basePath + 'NewControllerImage/heater_bg_offline.png'}} /> 
            <Image style={[styles.backgroundImage,{'position':'absolute',width:screenWidth,height:screenHeight}]}
                source={{isStatic:!MHPluginSDK.devMode, uri:MHPluginSDK.basePath + 'NewControllerImage/heater_box_offline.png'}}/> 
        </View>
      );
  }


  onPressButton(event) {

    this.refs.navigator.show();
    var key_id = undefined;
    this.state.allKeys.forEach(element => {
      
        if(element.name==event){
          key_id = element.id;
        }
    });
    console.log('key_id:'+key_id +'\n'+'event:'+event);

    if(key_id!=undefined){
    
      var self = this;
      var param  = {
        "did":this.state.did,
        controller_id:this.state.remoterID,
        "key_id": key_id
      };
      MIRequestServer.Post_SendkeyCode(param,
        function(response){
            if((response.code !=undefined) &&(response.code==0)){
              // self.refs.toast.showMessage(response.message);
            }
            
        },function(error){
            console.log('Post_SendkeyCode-error---'+JSON.stringify(error));
        });  
    }

  }

  isButtonEnabled(event) {

      var key_id = undefined;
      this.state.allKeys.forEach(element => {
          if(element.name==event){
            key_id = element.id;
          }
      });

      if(key_id!=undefined){
        return true;
      }else{
        return false;
      }
    
  }
  
  componentWillMount() {
     /*后期改成 controller_id */
     var param  = {"did": this.state.did};
     var self = this;
 
     MIRequestServer.Post_GetAllRemoterkey(param,
       function(response){
           if((response.result.keys)&&(response.result.keys!=undefined)){
             var allKeys = response.result.keys;
            //  self.deal_ExtendViewData(allKeys);
             self.setState({ allKeys:allKeys});
           }
 
       },function(error){
          self.refs.toast.showMessage(LocalizedStrings.ir_toast_get_data_failed);
       });  
  }

  deal_ExtendViewData(allKeys){

      var bindKeyArray =['POWER','TEMP+','TEMP-','MODE',];
      var NewAllKeys = allKeys.slice(0);
      var deleteArray =[];
      NewAllKeys.forEach((element,index) => {
          var isEqual = false;
          var sourceName = element.name;

          for(let i=0;i<bindKeyArray.length;i++){
              var key = bindKeyArray[i];
              if(key ==sourceName ){
                isEqual = true; 
                deleteArray.push(element);

              }
          }
      }); 
      deleteArray.forEach((element) => {

        let index = NewAllKeys.indexOf(element);      
        NewAllKeys.splice(index,1);
      });
      if(NewAllKeys.length>0){
        this.enableExtendButton =true;
      }
      return NewAllKeys;  
  }

  isEnableExtendButton(){
    return this.enableExtendButton;
  }

  onPressExtendButton(){
    this.setState({isShow:true});    
  }

  onPressDissmissButton(){
    this.setState({isShow:false});
  }

  componentWillUnmount() {

  }
}

const directionHeight = screenWidth*0.8;
const sectionHeght =( screenHeight - NavigatorBarHeight -300)/3.0;
const sectionWidth=  screenWidth;
const sectionLineWidth=  screenWidth;
const ButtonViewWidth=  screenWidth/3.0;
const ThreeButtonViewWidth=  screenWidth/3.0;
const FourButtonViewWidth =  screenWidth/4.0;
const contentBottomViewMarginTop = (screenHeight-NavigatorBarHeight -SafeBottomHeight) -(100+sectionHeght);

var styles = StyleSheet.create({

  containerAll: {
    flex:1,
  },
  contentBgView:{
    marginTop:0,
    height:screenHeight-NavigatorBarHeight,
  },
  contentBottomView:{
    marginBottom:50,
    marginTop:contentBottomViewMarginTop,
  },
  backgroundImage:{
    width:screenWidth,
    height:screenHeight,
  },
  directionView:{
    marginTop:NavigatorBarHeight,
    backgroundColor:'rgb(238,238,238)',
  },
  fourButtonView:{
    flexDirection:'row',
    width:FourButtonViewWidth,
    height:sectionHeght,
    alignItems:'center',
    alignSelf:'center',
    justifyContent:'center',
  },
  directionContentView:{
    width:320,
    height:320,
    flexDirection:'column',
    justifyContent:'center',
    alignItems:'center',
    alignSelf:'center',
    marginBottom:0,
  },
  image:{
    width:320,
    height:240,
    top:15,
    marginBottom:0,
  },
  lineSpace:{
    width:0.5,
    height:sectionHeght,
    backgroundColor:'#C0C0C0',
  },
  spaceLineColumn:{
    width:0.5,
    height:sectionHeght,
    backgroundColor:'#C0C0C0',
  },
  spaceLineRow:{
    height:0.5,
    width:sectionLineWidth,
    alignItems:'center',
    alignSelf:'center',
    backgroundColor:'#C0C0C0',
  },
  middleButtonView:{
    flexDirection:'row',
    width:sectionWidth,
    height:sectionHeght,
    alignItems:'center',
    justifyContent:'center',
  },
  ButtonView:{
    flexDirection:'row',
    width:ButtonViewWidth,
    height:sectionHeght,
    alignItems:'center',
    alignSelf:'center',
    justifyContent:'center',
  },
  threeButtonView:{
    flexDirection:'row',
    width:ThreeButtonViewWidth,
    height:sectionHeght,
    alignItems:'center',
    alignSelf:'center',
    justifyContent:'center',
  },

});

const KEY_OF_MAINPAGE = 'HeaterBoxVC';

// 打开更多菜单
var openMorePage = function (navigator) {
  navigator.push(Setting.route);
};

// 每个页面export自己的route
var route = {
  key: KEY_OF_MAINPAGE,
  title: MHPluginSDK.deviceName,
  component: HeaterBoxVC,
  navLeftButtonStyle: {
    tintColor:'#ffffff',
  },
  navTitleStyle: {
    color:'#ffffff',
  },
  navBarStyle: {
    backgroundColor:'transparent',
  },
  isNavigationBarHidden: false,
  renderNavLeftComponent: function(route, navigator, index, navState) {
    return (
      <View style={{left:0, width:29+15*2, height:APPBAR_HEIGHT, justifyContent:'center', alignItems:'center'}}>
        <ImageButton
          source={{uri:MHPluginSDK.uriNaviBackButtonImage, scale:PixelRatio.get()}}
          onPress={() => {
                if (index === 0) {
                  MHPluginSDK.closeCurrentPage();
                } else {
                  navigator.pop();
                }
          }}
          style={[{width:29, height:29, tintColor: '#ffffff'}]}
        />
      </View>
    );
  },
  renderNavRightComponent: function(route, navigator, index, navState) {
    if ((MHPluginSDK.userId == MHPluginSDK.ownerId)&&(index==0)) // 非分享设备
    {
      return (
        <View style={{left:0, width:29+15*2, height:APPBAR_HEIGHT, justifyContent:'center', alignItems:'center'}}>
          <ImageButton
            source={{uri:MHPluginSDK.uriNaviMoreButtonImage, scale:PixelRatio.get()}}
            onPress={() => {
              openMorePage(navigator);
            }}
            style={[{width:29, height:29, tintColor: '#ffffff'}]}
          />
        </View>
      );
    }
    else {
      return null;
    }
  },
}

module.exports = {
  component: HeaterBoxVC,
  route: route,
}
