// mh-infrared-gateway/Main/AirBoxSceneMain.js
'use strict';

var React = require('react-native');

var {
  AppRegistry,
  StyleSheet,
  Text,
  TouchableHighlight,
  Image,
  View,
  TextInput,
  PixelRatio,
  StatusBar,
  TouchableOpacity,
  Platform,
  ListView,
  Dimensions,
  DeviceEventEmitter,
} = React;

import { NativeModules } from 'react-native';
var MHPluginSDK = NativeModules.MHPluginSDK;
var ImageButton = require('../CommonModules/ImageButton');
var MIRequestServer = require('../MainV2/MIRequestServer/MIRequestServer').MIRequestServer;
var AirBoxNextSceneMain  =require('./AirBoxNextSceneMain');
var LocalizedStrings = require('../Main/MHLocalizableString.js').string;

var screenWidth = Dimensions.get('window').width;
var screenHeight = Dimensions.get('window').height;
const APPBAR_HEIGHT = Platform.OS === 'ios' ? 44 : 56;
const NavigatorBarHeight = (screenHeight==812)?(88):( APPBAR_HEIGHT +20);
const APP_MarginBottom =(screenHeight==812)?(34):(0);
const SafeBottomHeight =  (screenHeight==812)?(17):(0);

class AirBoxSceneMain extends React.Component {
  constructor(props, context) {
    super(props, context);

    this.state = {
      did: MHPluginSDK.deviceId,
      model: MHPluginSDK.deviceModel,
      apiLevel: MHPluginSDK.apiLevel,
      basePath: MHPluginSDK.basePath,
      devMode: MHPluginSDK.devMode,
      requestStatus: false,
      allKeys:[],
    };
  }

  componentWillMount() {

    var allKeys =[
    {"id":1,"name":"power_on","display_name":LocalizedStrings.ir_trun_on_air},
    {"id":2,"name":"power_off","display_name":LocalizedStrings.ir_trun_off_air},
    {"id":3,"name":"POWER","display_name":LocalizedStrings.ir_trunon_air_special}];
    this.setState({allKeys:allKeys});
  }

  componentDidMount() {
    console.log('Trigger: '+JSON.stringify(MHPluginSDK.extraInfo.trigger));
    console.log('Action: '+JSON.stringify(MHPluginSDK.extraInfo.action));
    console.log('Payload: '+JSON.stringify(MHPluginSDK.extraInfo.payload));
  }

  triggerRefreshUI() {
    var dataSource = new ListView.DataSource({rowHasChanged: (r1, r2)=>r1!==r2});
    this.state.dataSource = dataSource.cloneWithRows(this.state.allKeys);
  }

  render() {

    this.triggerRefreshUI();
    return (
      <View style={styles.containerAll} >
        <StatusBar />
        <View style={styles.containerView}>
          <ListView style={styles.remoterList} 
              enableEmptySections={true} 
              dataSource={this.state.dataSource} 
              renderRow={this._renderRow.bind(this)} 
              automaticallyAdjustContentInsets={false} />
          </View>
      </View>
    );
  }

   /* 列表一行，用来展示一个遥控器*/ 
   _renderRow(rowData,sectionID,rowID) {
    /*
          {"id":1,"name":"POWER","display_name":"电源"}
    */
    rowData.sectionID = sectionID;
    rowData.rowID = rowID;
    return (
          <TouchableHighlight underlayColor='#C0C0C0' onPress={() => this._pressRow(rowData)} >
            <View>
              <View style={styles.remoterListRowCell}>
                <Text style={styles.remoterName}>{rowData.display_name}</Text>
              </View>
            <View style={styles.separator}></View>
            </View>
          </TouchableHighlight>
    );
  }

  /* 点击一行 */
  _pressRow(rowData){
    

   
    if(rowData.rowID==0){
        
      
        var action=MHPluginSDK.extraInfo.action;
        action.deviceAbbr="红外遥控器";
        action.keyName=LocalizedStrings.ir_trun_on_air;
        var params ={'keys':['power_on']};
        action.payload.value=params;
        MHPluginSDK.finishCustomSceneSetupWithAction(action);
    }else if(rowData.rowID==1){

        var action=MHPluginSDK.extraInfo.action;
        action.deviceAbbr="红外遥控器";  
        action.keyName=LocalizedStrings.ir_trun_off_air;
        var params ={'keys':['power_off']};
        action.payload.value=params;
        MHPluginSDK.finishCustomSceneSetupWithAction(action);
    }else{

        this.props.navigator.push({
          ...AirBoxNextSceneMain.route,
          passProps: {}
        });
    }

  }

}

var styles = StyleSheet.create({
  containerAll: {
    flex: 1,
    flexDirection: 'column',
    backgroundColor: '#ffffff',
    marginTop: 0,
  },
  containerView:{
    marginTop:NavigatorBarHeight,
    height:screenHeight-NavigatorBarHeight-SafeBottomHeight,
    width:screenWidth,
  },
  remoterList:{
    marginTop:0,
    height:screenHeight-NavigatorBarHeight-SafeBottomHeight,
    width:screenWidth,
  },
  separator: {
    height: 0.5,
    backgroundColor: 'rgba(0, 0, 0, 0.2)',
    marginLeft: 20,
    marginRight: 20
  },
  remoterListRowCell: {
    flexDirection: 'row',
    padding: 20,
    height: 60,
  },
  remoterImage: {
    alignSelf: 'center',
    width: 46,
    height: 46,
  },
  remoterName: {
    fontSize: 15,
    opacity: 0.9,
    alignSelf: 'center',
    color:'#000000',
    marginLeft: 0,
  },
});

const KEY_OF_SCENEMAIN = 'AirBoxSceneMain';

// 每个页面export自己的route
var route = {
  key: KEY_OF_SCENEMAIN,
  title: '自定义场景',
  component: AirBoxSceneMain,
  navLeftButtonStyle: {
    tintColor:'#808080',
  },
  navTitleStyle: {
    color:'#808080',
  },
  navBarStyle: {
    backgroundColor:'transparent',
  },
  renderNavLeftComponent(route, navigator, index, navState) {
    return (<View style={{left:0, width:29+15*2, height:APPBAR_HEIGHT, justifyContent:'center', alignItems:'center'}}>
      <ImageButton
        source={{uri:MHPluginSDK.uriNaviBackButtonImage, scale:PixelRatio.get()}}
        onPress={() => {
          if (index === 0) {
            MHPluginSDK.finishCustomSceneSetup(MHPluginSDK.extraInfo.payload);
            MHPluginSDK.closeCurrentPage();
          } else {
            navigator.pop();
          }
        }}
        style={[{width:29, height:29, tintColor: '#000000'}, route.navLeftButtonStyle]}
      />
    </View>);
  },
  isNavigationBarHidden: false,
}

module.exports = {
  component: AirBoxSceneMain,
  route: route,
}
