// gl-react/lib/AnimatedData.js
"use strict";

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var isAnimated = require("./isAnimated");

// At the moment, need to dup some things from RN Animated

var Animated = function () {
  function Animated() {
    _classCallCheck(this, Animated);
  }

  _createClass(Animated, [{
    key: "__attach",
    value: function __attach() {}
  }, {
    key: "__detach",
    value: function __detach() {}
  }, {
    key: "__getValue",
    value: function __getValue() {}
  }, {
    key: "__getAnimatedValue",
    value: function __getAnimatedValue() {
      return this.__getValue();
    }
  }, {
    key: "__addChild",
    value: function __addChild() {}
  }, {
    key: "__removeChild",
    value: function __removeChild() {}
  }, {
    key: "__getChildren",
    value: function __getChildren() {
      return [];
    }
  }]);

  return Animated;
}();

var AnimatedWithChildren = function (_Animated) {
  _inherits(AnimatedWithChildren, _Animated);

  function AnimatedWithChildren() {
    _classCallCheck(this, AnimatedWithChildren);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(AnimatedWithChildren).call(this));

    _this._children = [];
    return _this;
  }

  _createClass(AnimatedWithChildren, [{
    key: "__addChild",
    value: function __addChild(child) {
      if (this._children.length === 0) {
        this.__attach();
      }
      this._children.push(child);
    }
  }, {
    key: "__removeChild",
    value: function __removeChild(child) {
      var index = this._children.indexOf(child);
      if (index === -1) {
        console.warn("Trying to remove a child that doesn't exist");
        return;
      }
      this._children.splice(index, 1);
      if (this._children.length === 0) {
        this.__detach();
      }
    }
  }, {
    key: "__getChildren",
    value: function __getChildren() {
      return this._children;
    }
  }]);

  return AnimatedWithChildren;
}(Animated);

// Animated over the GL Data uniforms object


var AnimatedUniforms = function (_AnimatedWithChildren) {
  _inherits(AnimatedUniforms, _AnimatedWithChildren);

  function AnimatedUniforms(uniforms) {
    _classCallCheck(this, AnimatedUniforms);

    var _this2 = _possibleConstructorReturn(this, Object.getPrototypeOf(AnimatedUniforms).call(this));

    _this2._uniforms = uniforms;
    _this2.__attach();
    return _this2;
  }

  _createClass(AnimatedUniforms, [{
    key: "__getValue",
    value: function __getValue() {
      var u = {};
      var uniforms = this._uniforms;
      for (var key in uniforms) {
        var value = uniforms[key];
        if (value instanceof Array) {
          var arr = [];
          for (var i = 0; i < value.length; i++) {
            var v = value[i];
            arr[i] = isAnimated(v) ? v.__getValue() : v;
          }
          u[key] = arr;
        } else if (isAnimated(value)) {
          u[key] = value.__getValue();
        } else {
          u[key] = value;
        }
      }
      return u;
    }
  }, {
    key: "__attach",
    value: function __attach() {
      var uniforms = this._uniforms;
      for (var key in uniforms) {
        var value = uniforms[key];
        if (value instanceof Array) {
          for (var i = 0; i < value.length; i++) {
            var v = value[i];
            if (isAnimated(v)) {
              v.__addChild(this);
            }
          }
        } else if (isAnimated(value)) {
          value.__addChild(this);
        }
      }
    }
  }, {
    key: "__detach",
    value: function __detach() {
      var uniforms = this._uniforms;
      for (var key in uniforms) {
        var value = uniforms[key];
        if (value instanceof Array) {
          for (var i = 0; i < value.length; i++) {
            var v = value[i];
            if (isAnimated(v)) {
              v.__removeChild(this);
            }
          }
        } else if (isAnimated(value)) {
          value.__removeChild(this);
        }
      }
    }
  }]);

  return AnimatedUniforms;
}(AnimatedWithChildren);

// Animated over a GL Data


var AnimatedData = function (_AnimatedWithChildren2) {
  _inherits(AnimatedData, _AnimatedWithChildren2);

  function AnimatedData(data, callback) {
    _classCallCheck(this, AnimatedData);

    var _this3 = _possibleConstructorReturn(this, Object.getPrototypeOf(AnimatedData).call(this));

    _this3._data = _extends({}, data, {
      contextChildren: data.contextChildren.map(function (d) {
        return new AnimatedData(d);
      }),
      children: data.children.map(function (d) {
        return new AnimatedData(d);
      }),
      uniforms: new AnimatedUniforms(data.uniforms)
    });
    if (callback) _this3.update = callback;
    _this3.__attach();
    return _this3;
  }

  _createClass(AnimatedData, [{
    key: "__getValue",
    value: function __getValue() {
      var _data = this._data;

      var data = _objectWithoutProperties(_data, []);

      var contextChildren = _data.contextChildren;
      var width = _data.width;
      var height = _data.height;
      var children = _data.children;
      var uniforms = _data.uniforms;

      data.width = isAnimated(width) ? width.__getValue() : width;
      data.height = isAnimated(height) ? height.__getValue() : height;
      data.contextChildren = contextChildren.map(function (c) {
        return c.__getValue();
      });
      data.children = children.map(function (c) {
        return c.__getValue();
      });
      data.uniforms = uniforms.__getValue();
      return data;
    }
  }, {
    key: "__attach",
    value: function __attach() {
      var _this4 = this;

      var _data2 = this._data;
      var contextChildren = _data2.contextChildren;
      var children = _data2.children;
      var uniforms = _data2.uniforms;
      var width = _data2.width;
      var height = _data2.height;

      if (isAnimated(width)) width.__addChild(this);
      if (isAnimated(height)) height.__addChild(this);
      contextChildren.forEach(function (c) {
        return c.__addChild(_this4);
      });
      children.forEach(function (c) {
        return c.__addChild(_this4);
      });
      uniforms.__addChild(this);
    }
  }, {
    key: "__detach",
    value: function __detach() {
      var _this5 = this;

      var _data3 = this._data;
      var contextChildren = _data3.contextChildren;
      var children = _data3.children;
      var uniforms = _data3.uniforms;
      var width = _data3.width;
      var height = _data3.height;

      if (isAnimated(width)) width.__removeChild(this);
      if (isAnimated(height)) height.__removeChild(this);
      contextChildren.forEach(function (c) {
        return c.__removeChild(_this5);
      });
      children.forEach(function (c) {
        return c.__removeChild(_this5);
      });
      uniforms.__removeChild(this);
    }
  }]);

  return AnimatedData;
}(AnimatedWithChildren);

module.exports = AnimatedData;
