// gl-react/lib/createSurface.js
"use strict";

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var React = require("react");
var Component = React.Component;
var PropTypes = React.PropTypes;

var invariant = require("invariant");

var _require = require("./data");

var fill = _require.fill;
var resolve = _require.resolve;
var build = _require.build;

var Shaders = require("./Shaders");
var Node = require("./Node");
var postShader = require("./postShader");
var findGLNodeInGLComponentChildren = require("./data/findGLNodeInGLComponentChildren");
var invariantStrictPositive = require("./data/invariantStrictPositive");
var AnimatedData = require("./AnimatedData");
var runtime = require("./runtime");

var _glSurfaceId = 1;

function logResult(data, contentsVDOM) {
  if (typeof console !== "undefined" && console.debug // eslint-disable-line
  ) {
      console.debug("GL.Surface rendered with", data, contentsVDOM); // eslint-disable-line no-console
    }
}

module.exports = function (renderVcontainer, renderVcontent, renderVGL, getPixelRatio) {
  var _getGLCanvas = arguments.length <= 4 || arguments[4] === undefined ? function (glSurface) {
    return glSurface.refs.canvas;
  } : arguments[4];

  var GLSurface = function (_Component) {
    _inherits(GLSurface, _Component);

    function GLSurface(props, context) {
      _classCallCheck(this, GLSurface);

      var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(GLSurface).call(this, props, context));

      _this._renderId = 0;
      _this._id = _glSurfaceId++;
      return _this;
    }

    _createClass(GLSurface, [{
      key: "componentWillMount",
      value: function componentWillMount() {
        Shaders._onSurfaceWillMount(this._id);
        this._build(this.props);
        this._attach();
      }
    }, {
      key: "componentWillUnmount",
      value: function componentWillUnmount() {
        this._renderId = 0;
        Shaders._onSurfaceWillUnmount(this._id);
        this._dataAnimated && this._dataAnimated.__detach();
      }
    }, {
      key: "componentWillReceiveProps",
      value: function componentWillReceiveProps(nextProps) {
        this._build(nextProps);
        this._attach();
      }
    }, {
      key: "_build",
      value: function _build(props) {
        var _this2 = this;

        var id = this._id;
        var renderId = ++this._renderId;
        var width = props.width;
        var height = props.height;
        var pixelRatioProps = props.pixelRatio;
        var children = props.children;
        var debug = props.debug;
        var preload = props.preload;


        invariant(children, "GL.Surface must have in children a GL.Node or a GL Component");

        var decorateOnShaderCompile = function decorateOnShaderCompile(onShaderCompile) {
          return onShaderCompile && // only decorated if onShaderCompile is defined
          function (error, result) {
            return renderId === _this2._renderId && // it's outdated. skip the callback call
            onShaderCompile(error, result);
          };
        }; // it's current. propagate the call

        var pixelRatio = pixelRatioProps || getPixelRatio(props);

        invariantStrictPositive(pixelRatio, "GL.Surface: pixelRatio prop");

        var surfaceContext = {
          width: width,
          height: height,
          pixelRatio: pixelRatio
        };

        var glNode = findGLNodeInGLComponentChildren(React.createElement(Node, _extends({
          shader: postShader
        }, surfaceContext, {
          uniforms: { t: children }
        })), surfaceContext);

        invariant(glNode && glNode.childGLNode, "GL.Surface must have in children a GL.Node or a GL Component");

        var via = glNode.via;
        var childGLNode = glNode.childGLNode;
        var context = glNode.context;


        var resolved = void 0;
        try {
          Shaders._beforeSurfaceBuild(id);
          resolved = resolve(fill(build(childGLNode, context, preload, via, id, decorateOnShaderCompile)));
        } catch (e) {
          throw e;
        } finally {
          Shaders._afterSurfaceBuild(id);
        }

        this._resolved = resolved;
        this._pixelRatio = pixelRatio;

        if (debug) logResult(resolved.data, resolved.contentsVDOM);
      }
    }, {
      key: "_attach",
      value: function _attach() {
        var _this3 = this;

        var oldDataAnimated = this._dataAnimated;
        var callback = function callback() {
          var canvas = _this3.getGLCanvas();
          if (!canvas) return;
          if (canvas.setNativeProps) {
            var data = _this3._dataAnimated.__getValue();
            canvas.setNativeProps({ data: data });
          } else {
            _this3.forceUpdate();
          }
        };
        this._dataAnimated = new AnimatedData(this._resolved.data, callback);

        oldDataAnimated && oldDataAnimated.__detach();
      }
    }, {
      key: "getGLCanvas",
      value: function getGLCanvas() {
        return _getGLCanvas(this);
      }
    }, {
      key: "captureFrame",
      value: function captureFrame() {
        var c = this.getGLCanvas();
        invariant(c, "c is '%s'. Is the component unmounted?", c);
        invariant(c.captureFrame, "captureFrame() should be implemented by GLCanvas");
        return c.captureFrame.apply(c, arguments);
      }
    }, {
      key: "render",
      value: function render() {
        var renderId = this._renderId;
        var _resolved = this._resolved;
        var contentsVDOM = _resolved.contentsVDOM;
        var imagesToPreload = _resolved.imagesToPreload;

        var data = this._dataAnimated.__getValue();
        var pixelRatio = this._pixelRatio;
        var props = this.props;
        var children = props.children;
        var debug = props.debug;
        var preload = props.preload;
        var // eslint-disable-line no-unused-vars
        style = props.style;
        var width = props.width;
        var height = props.height;
        var backgroundColor = props.backgroundColor;
        var visibleContent = props.visibleContent;
        var eventsThrough = props.eventsThrough;

        var restProps = _objectWithoutProperties(props, ["children", "debug", "preload", "style", "width", "height", "backgroundColor", "visibleContent", "eventsThrough"]);

        return renderVcontainer({ width: width, height: height, style: style, visibleContent: visibleContent, eventsThrough: eventsThrough }, contentsVDOM.map(function (vdom, i) {
          return renderVcontent(data.width, data.height, i, runtime.decorateVDOMContent(vdom), { visibleContent: visibleContent });
        }), renderVGL(_extends({}, restProps, { // eslint-disable-line no-undef
          style: { backgroundColor: backgroundColor },
          width: width,
          height: height,
          pixelRatio: pixelRatio,
          data: data,
          nbContentTextures: contentsVDOM.length,
          imagesToPreload: imagesToPreload,
          renderId: renderId,
          visibleContent: visibleContent,
          eventsThrough: eventsThrough
        })));
      }
    }]);

    return GLSurface;
  }(Component);

  GLSurface.displayName = "GL.Surface";

  GLSurface.propTypes = {
    width: PropTypes.any.isRequired,
    height: PropTypes.any.isRequired,
    backgroundColor: PropTypes.string,
    pixelRatio: PropTypes.number,
    children: PropTypes.element.isRequired,
    preload: PropTypes.bool,
    autoRedraw: PropTypes.bool,
    eventsThrough: PropTypes.bool,
    visibleContent: PropTypes.bool,
    onLoad: PropTypes.func,
    onProgress: PropTypes.func
  };

  GLSurface.defaultProps = {
    preload: false,
    autoRedraw: false,
    eventsThrough: false,
    visibleContent: false,
    backgroundColor: "#000"
  };

  return GLSurface;
};
