// mh-infrared-gateway/MainV2/MIRequestServer/AirStateModel.js
'use strict';

var React = require('react-native');

var {
  Component,
} = React;

let Temperature  =[
'ir2_air_conditioner_temp_num_0.png',
'ir2_air_conditioner_temp_num_1.png',
'ir2_air_conditioner_temp_num_2.png',
'ir2_air_conditioner_temp_num_3.png',
'ir2_air_conditioner_temp_num_4.png',
'ir2_air_conditioner_temp_num_5.png',
'ir2_air_conditioner_temp_num_6.png',
'ir2_air_conditioner_temp_num_7.png',
'ir2_air_conditioner_temp_num_8.png',
'ir2_air_conditioner_temp_num_9.png',
'',
];

let airMode  ={
 'auto':{
    'mode':'auto',
    'Title':'自动',
    'imageName':'air_show_auto.png'
  },
  'humidity': {
    'mode':'humidity',
    'Title':'除湿',
    'imageName':'air_show_wet.png'
  },
  'cool': {
    'mode':'cool',
    'Title':'制冷',
    'imageName':'air_show_cool.png'
  },
  'heat': {
    'mode':'heat',
    'Title':'制热',
    'imageName':'air_show_hot.png'
  },
  'winddirect':{
    'mode':'winddirect',
    'Title':'通风',
    'imageName':'air_show_scanWind.png'
  },
  'none': {
    'mode':'',
    'Title':'',
    'imageName':''
  },
};

let windSpeed = {
  'auto':{
    'speed':'auto',
    'Title':'风速 自动',
    'imageName':'air_show_auto.png'
  },
  'low': {
    'speed':'low',
    'Title':'风速 低',
    'imageName':'air_show_speed1.png'
  },
  'middle': {
    'speed':'middle',
    'Title':'风速 中',
    'imageName':'air_show_speed2.png'
  },
  'high': {
    'speed':'high',
    'Title':'风速 高',
    'imageName':'air_show_speed3.png'
  },
  'none': {
    'speed':'',
    'Title':'',
    'imageName':''
  },

};

let windDirection = {
  'up': {
    'direction':'up',
    'Title':'风向上',
    'imageName':'irv2_air_conditioner_winddirect_up_v5.png'
  },
  'middle': {
    'direction':'middle',
    'Title':'风向中',
    'imageName':'irv2_air_conditioner_winddirect_middle_v5.png'
  },
  'down': {
    'direction':'down',
    'Title':'风向下',
    'imageName':'irv2_air_conditioner_winddirect_down_v5.png'
  },
  'none': {
    'direction':'',
    'Title':'',
    'imageName':''
  },
};

let scanWind = {
 'auto': {
    'scanMode':'auto',
    'Title':'自动风向',
    'imageName':'irv2_air_conditioner_btn_windspread_v5.png'
  },
 'default': {
    'scanMode':'default',
    'Title':'手动风向',
    'imageName':'irv2_air_conditioner_btn_windspread_default_v5.png'
  },
  'none': {
    'scanMode':'',
    'Title':'',
    'imageName':''
  },
};

let sleepMode = {
  'sleep': {
    'scanMode':'sleep',
    'Title':'睡眠',
    'imageName':'air_show_sleep.png'
  },
  'none': {
    'scanMode':'',
    'Title':'',
    'imageName':''
  },
};

module.exports.Temperature = Temperature;
module.exports.airMode = airMode;
module.exports.windSpeed = windSpeed;
module.exports.windDirection = windDirection;
module.exports.scanWind = scanWind;
module.exports.sleepMode = sleepMode;

// export { Temperature, airMode , windSpeed, windDirection, scanWind, sleepMode}
